### PR TITLE
add support for xsdb runner

### PR DESCRIPTION
**Pull request by kedareswararao**
This pull request adds support for xsdb runner

### CODE REVIEW

File: scripts/west_commands/runners/__init__.py [hunk 1]
```diff
@@ -60,6 +60,7 @@ def _import_runner_module(runner_name):
     'trace32',
     'uf2',
     'xtensa',
+    'xsdb',
```
Comment (pdgendt) [inline scripts/west_commands/runners/__init__.py:63]: Please read the line below this one ;)

Comment (pdgendt) [reply scripts/west_commands/runners/__init__.py:63]: Since you're touching this file, wrap these entries with:

```
# zephyr-keep-sorted-start
...
# zephyr-keep-sorted-stop
```

Comment (kedareswararao) [reply scripts/west_commands/runners/__init__.py:63]: Sure will fix...

File: boards/amd/kv260_r5/board.cmake
Comment (henrikbrixandersen) [inline boards/amd/kv260_r5/board.cmake:1]: I would suggest creating a `boards/common/xsdb.board.cmake` for setting the defaults and include this in the `board.cmake`. Please see the in-tree examples of how this is done (e.g. `boards/common/jlink.board.cmake).

File: boards/amd/kv260_r5/support/xsdb.cfg [hunk 2]
```diff
@@ -0,0 +1,25 @@
+# Copyright (c) 2024 Advanced Micro Devices, Inc.
+#
+# SPDX-License-Identifier: Apache-2.0
+# Author: Appana Durga Kedareswara rao
```
Comment (henrikbrixandersen) [inline boards/amd/kv260_r5/support/xsdb.cfg:4]: We usually don't list authors.

File: scripts/west_commands/runners/xsdb.py
Comment (henrikbrixandersen) [inline scripts/west_commands/runners/xsdb.py:1]: Please add a test suite under `scripts/west_commands/tests/`.

File: scripts/west_commands/runners/xsdb.py [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+        super(XSDBBinaryRunner, self).__init__(cfg)
+        self.elf_file = cfg.elf_file
+        if not config:
+            cfgfile_path = os.path.join(cfg.board_dir, 'support')
+            default = os.path.join(cfgfile_path, 'xsdb.cfg')
+            board_name = os.path.basename(cfg.board_dir)
+            if 'mbv32' in board_name and bitstream is None:
+                raise ValueError('Valid bitstream is required for mbv32 platform')
+            if 'kv260_r5' in board_name and fsbl is None:
+                raise ValueError('Valid fsbl elf is required for kv260_r5 platform')
```
Comment (henrikbrixandersen) [inline scripts/west_commands/runners/xsdb.py:25]: A generic runner should not contain board-specific code.

**Review by henrikbrixandersen — CHANGES_REQUESTED** [review]

File: scripts/west_commands/runners/__init__.py (hunk 1 above)
Comment (pdgendt) [reply scripts/west_commands/runners/__init__.py:63]: It appears the list wasn't sorted to begin with, I've created #80061 to fix this.

File: scripts/west_commands/runners/xsdb.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (kedareswararao) [reply scripts/west_commands/runners/xsdb.py:25]: Sure will fix

File: boards/amd/kv260_r5/board.cmake [hunk 4]
```diff
@@ -1,3 +1,8 @@
 # Copyright (c) 2022 Linaro.
 #
 # SPDX-License-Identifier: Apache-2.0
+board_set_runner(FLASH xsdb)
+board_set_debugger_ifnset(xsdb)
+board_set_flasher_ifnset(xsdb)
+board_finalize_runner_args(xsdb)
+set(FLASH_RUNNER xsdb)
```
[the remark this answers is not in the document]
Comment (kedareswararao) [reply boards/amd/kv260_r5/board.cmake:8]: Sure will move it to xsdb.board.cmake and will include it in the board.cmake

File: boards/amd/kv260_r5/board.cmake
Comment (kedareswararao) [reply boards/amd/kv260_r5/board.cmake:1]: sure

File: boards/amd/kv260_r5/support/xsdb.cfg (hunk 2 above)
Comment (kedareswararao) [reply boards/amd/kv260_r5/support/xsdb.cfg:4]: Okay will remove the Author tag

File: scripts/west_commands/runners/xsdb.py
Comment (kedareswararao) [reply scripts/west_commands/runners/xsdb.py:1]: sure will add

File: scripts/west_commands/runners/xsdb.py (hunk 3 above)
Comment (kedareswararao) [reply scripts/west_commands/runners/xsdb.py:25]: okay will remove board specific logic

File: scripts/west_commands/runners/xsdb.py [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+    def __init__(self, cfg: RunnerConfig, config=None, bitstream=None,
+            fsbl=None):
+        super(XSDBBinaryRunner, self).__init__(cfg)
+        self.elf_file = cfg.elf_file
+        if not config:
+            cfgfile_path = os.path.join(cfg.board_dir, 'support')
+            default = os.path.join(cfgfile_path, 'xsdb.cfg')
+            board_name = os.path.basename(cfg.board_dir)
+            if os.path.exists(default):
+                config= default
```
Comment (pdgendt) [inline scripts/west_commands/runners/xsdb.py:22]: Nit
```suggestion
                config = default
```

File: scripts/west_commands/tests/test_xsdb.py [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+    arg_namespace = parser.parse_args(args)
+
+    # Mock the default config path
+    with patch("runners.xsdb.os.path.join", return_value="default_cfg_path"):
+        runner = XSDBBinaryRunner.create(runner_config, arg_namespace)
+
+    runner.do_run("flash")
+
+    assert check_call.call_args_list == [call(tc["expected_cmd"])]
+
```
Comment (pdgendt) [inline scripts/west_commands/tests/test_xsdb.py:80]: Nit, remove extra newline:
```suggestion
```

**Comment by pdgendt** [comment]
Fix compliance issues too please

**Comment by kedareswararao** [comment]
> Fix compliance issues too please

@pdgendt: Sure could you please help me on this is there tests that can be ran for compliance testing? currently I am testing the runner using twister and west flash commands and using pytest validated the test_xsdb.py.

File: scripts/west_commands/runners/xsdb.py (hunk 5 above)
Comment (kedareswararao) [reply scripts/west_commands/runners/xsdb.py:22]: sure will fix

**Comment by pdgendt** [comment]
> @pdgendt: Sure could you please help me on this is there tests that can be ran for compliance testing? currently I am testing the runner using twister and west flash commands and using pytest validated the test_xsdb.py.

Check the failed actions for more details:
https://github.com/zephyrproject-rtos/zephyr/actions/runs/11416910747/job/31808580843?pr=80046
https://github.com/zephyrproject-rtos/zephyr/actions/runs/11416910743/job/31808580957?pr=80046

**Comment by pdgendt** [comment]
The last compliance issue will be fixed with #80061 rebase once that is merged.

**Comment by kedareswararao** [comment]
> The last compliance issue will be fixed with #80061 rebase once that is merged.

Seen your pull request that's why didn't done changes on the __init__.py file will rebase the changes once above pull request got merged.

File: boards/amd/kv260_r5/board.cmake [hunk 7]
```diff
@@ -1,3 +1,4 @@
 # Copyright (c) 2022 Linaro.
 #
 # SPDX-License-Identifier: Apache-2.0
+include(${ZEPHYR_BASE}/boards/common/xsdb.board.cmake)
```
Comment (pdgendt) [inline boards/amd/kv260_r5/board.cmake:4]: Shouldn't this get an `fsbl` argument? Or am I missing something?

Comment (pdgendt) [reply boards/amd/kv260_r5/board.cmake:4]: Is the `fsbl.elf` file generated from the build? Ideally the necessary files are passed by default to the runner, so users can simply call `west flash` without the need to add manual arguments.

Comment (pdgendt) [reply boards/amd/kv260_r5/board.cmake:4]: For example something like:

```cmake
board_runner_args(xsdb "--fsbl=${PROJECT_BINARY_DIR}/fsbl.elf") 
include(${ZEPHYR_BASE}/boards/common/xsdb.board.cmake)
```

Comment (kedareswararao) [reply boards/amd/kv260_r5/board.cmake:4]: Thanks for the example, will update and document the usage in the commit message.

Comment (pdgendt) [reply boards/amd/kv260_r5/board.cmake:4]: > fsbl.elf is a bootloader binary built using AMD tools it won't be built using zephyr sources and this file is required inorder to initialize the hardware before loading the zephyr binaries.

Perhaps this should become a blob then?

Comment (kedareswararao) [reply boards/amd/kv260_r5/board.cmake:4]: Not sure about blob but i think we should use file option in the RunnerConfig for this instead of introducing a new option fsbl and bitstream and pass that file from board_runner_args like you suggested above.  
file: Optional[str]             # binary file path (provided by the user), or None
let me know if this approach is fine for you or not, if you are okay will update the code accordingly.

Comment (kedareswararao) [reply boards/amd/kv260_r5/board.cmake:5]: it's a manual action okay let me explore how west blobs fetch works

Comment (henrikbrixandersen) [reply boards/amd/kv260_r5/board.cmake:4]: The First Stage Boot Loader (fsbl.elf) is a design-specific (note: not only board-specific, but specific to the given PS/PL configuration used for a given design). It is is built outside of the Zephyr build environment and - in my experience - seldomly needs to be rebuilt.

Letting the user provide the path to the fsbl.elf in the runner perfectly aligns with how we use this downstream. I don't see how this would fit with the `west blobs` feature.

Comment (kedareswararao) [reply boards/amd/kv260_r5/board.cmake:4]: Thanks @henrikbrixandersen for the inputs, @pdgendt existing/latest code aligns with the @henrikbrixandersen suggestion, if you are okay with the changes please approve the pull request

Regerds,
Kedar.

**Review by pdgendt — DISMISSED** [review]

File: boards/common/xsdb.board.cmake [hunk 8]
```diff
@@ -0,0 +1,8 @@
+# Copyright (c) 2024 Advanced Micro Devices, Inc.
+#
+# SPDX-License-Identifier: Apache-2.0
+board_set_runner(FLASH xsdb)
```
Comment (henrikbrixandersen) [inline boards/common/xsdb.board.cmake:4]: Why is this line needed? None of the other *.board.cmake files seem to set this.

File: scripts/west_commands/runners/xsdb.py [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+
+    @classmethod
+    def capabilities(cls):
+        return RunnerCaps(flash_addr=True)
+
+    @classmethod
+    def do_add_parser(cls, parser):
+        parser.add_argument('--config', help='if given, override default config file')
+        parser.add_argument('--bitstream', help='path to the bitstream file')
+        parser.add_argument('--fsbl', help='path to the r5 fsbl elf file')
```
Comment (henrikbrixandersen) [inline scripts/west_commands/runners/xsdb.py:38]: I'd suggest removing "r5" from here, as this should work with e.g. Zynq 7000 series as well?

```suggestion
        parser.add_argument('--fsbl', help='path to the fsbl.elf file')
```

File: scripts/west_commands/runners/xsdb.py [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+        return XSDBBinaryRunner(cfg, config=args.config,
+                bitstream=args.bitstream, fsbl=args.fsbl)
+
+    def do_run(self, command, **kwargs):
+        if self.bitstream:
+            cmd = ['xsdb', self.xsdb_cfg_file, self.elf_file, self.bitstream]
+        elif self.fsbl:
+            cmd = ['xsdb', self.xsdb_cfg_file, self.elf_file, self.fsbl]
+        else:
+            cmd = ['xsdb', self.xsdb_cfg_file]
```
Comment (henrikbrixandersen) [inline scripts/west_commands/runners/xsdb.py:53]: Are these really mutually exclusive? Either a bitstream or an FSBL, but never both?

**Review by henrikbrixandersen — CHANGES_REQUESTED** [review]
Looks good. A few minor observations/questions:

File: scripts/west_commands/runners/xsdb.py (hunk 9 above)
Comment (kedareswararao) [reply scripts/west_commands/runners/xsdb.py:38]: Sure will remove

File: scripts/west_commands/runners/xsdb.py (hunk 10 above)
Comment (kedareswararao) [reply scripts/west_commands/runners/xsdb.py:53]: They can be mutually inclusive i mean both can present in that case expecting users to take care of this use case by passing custom config file (else condition)

**Comment by henrikbrixandersen** [comment]
(also, please avoid rebasing on `main` unless to resolve a merge conflict or pull in a needed commit - this makes comparing changes between revisions much more difficult)

**Comment by kedareswararao** [comment]
> (also, please avoid rebasing on `main` unless to resolve a merge conflict or pull in a needed commit - this makes comparing changes between revisions much more difficult)

sure sorry for the noise

File: scripts/west_commands/runners/xsdb.py (hunk 10 above)
Comment (kedareswararao) [reply scripts/west_commands/runners/xsdb.py:53]: Sure will add a check and error out

File: boards/common/xsdb.board.cmake [hunk 11]
```diff
@@ -0,0 +1,7 @@
+# Copyright (c) 2024 Advanced Micro Devices, Inc.
+#
+# SPDX-License-Identifier: Apache-2.0
+board_set_debugger_ifnset(xsdb)
+board_set_flasher_ifnset(xsdb)
+board_finalize_runner_args(xsdb)
+set(FLASH_RUNNER xsdb)
```
Comment (henrikbrixandersen) [inline boards/common/xsdb.board.cmake:7]: What is the last line here doing?

**Review by henrikbrixandersen — CHANGES_REQUESTED** [review]
Looks good, thanks. One last question:

File: boards/common/xsdb.board.cmake (hunk 11 above)
Comment (kedareswararao) [reply boards/common/xsdb.board.cmake]: it's also a unnecessary line removed now thanks for the quick review and reply.

**Review by henrikbrixandersen — APPROVED** [review]

**Review by pdgendt — APPROVED** [review]

**Comment by kedareswararao** [comment]
@swinslow @carlescufi @mbolivar-ampere please review the pull request when you have sometime
